tests/test_replay.py

```diff
@@ -74,6 +74,10 @@ def test_replay_id(self):
         # we can parse it properly instead of erroring
         self.assertEqual(self._old_replayid_replay.replay_id, 1127598189)
 
+    def test_replay_hash(self):
+        for replay in self._replays:
+            self.assertEqual(Replay.calculate_replay_hash(replay), "b06ecaf5fc301545b8b23769b2a20451", "Replay hash is wrong")
```

Comment:
why not just compare to `replay.replay_hash` instead of hardcoding `b06ecaf5fc301545b8b23769b2a20451`

Reply:
with our settings of compressing replay events to LZMA, the replay hash in an original replay is wrong
unless we want to change it in the file, it needs to be hardcoded

Reply:
Isn't the whole point of this change to be able to match the hash produced by osu? We should make the changes necessary to make that happen, or if it's not feasible, then I don't think we should be including the ability to calculate replay hashes at all.

Reply:
In other words, the following needs to hold for all replays for this PR to have any value:

```python
Replay.calculate_replay_hash(replay) == replay.replay_hash
```

Reply:
> Isn't the whole point of this change to be able to match the hash produced by osu?

the original point was to generate a hash for osu! (or any other clients for osu) based on replay data, in case it checks for the hash itself

even though the current stable version doesn't care, i am not sure about osu!lazer and other clients that support replays

Reply:
But we don't want to just generate any hash, we want to generate the exact same hash that osu! is generating. I'm not even sure if this is possible since it's not documented how the osr hash is generated. If it's not possible for us to match osu's hash, then we shouldn't even try to calculate our own hash - we don't get to decide for ourselves what the replay hash should be. That's up to osu to specify and for us to follow.

Reply:
> But we don't want to just generate any hash, we want to generate the exact same hash that osu! is generating.

ah okay, i understand you
in that case, we would need to compress replay events into LZMA with the same settings as osu! does, since that's what it generates a hash from later on
i am not sure if it's possible to compress it in the same way again, considering we don't know exact settings used in osu! used for compressing

however, still, the point was to generate a hash, in case any client generates it for themselves based on compressed data in our packed replay and compares it to the one in the replay

Reply:
> however, still, the point was to generate a hash, in case any client generates it for themselves based on compressed data in our packed replay and compares it to the one in the replay

that would be good, but how do we know that an md5 hash of the replay data is going to return the hash we want? We need some way to check that we're using the correct hash function. Either by finding documentation that says osu uses the md5 of the replay data, or by verifying ourselves that 

```python
Replay.calculate_replay_hash(replay) == replay.replay_hash
```

for all replays, as above. We can't just guess that they're md5 hashing the replay data and hope we get lucky (unless you have a source that says that's what they're doing; it might be).

Maybe osu takes the timestamp of the replay into account for the hash, for instance, and not just the replay data. That's going to be the hard part of this pull request - figuring out exactly how to generate the replay hash (if it's possible at all). As you said, it might be as simple as changing our lzma compression algorithms.

Reply:
ah right, sorry, i forgot about the `(includes certain properties of the replay)` mention on osu! wiki.
well, we can't really do that much other than find information about it, if someone documented it, look into the source code of other replay writers, if some of them perhaps alter the hash in a correct way, or brute force our way and test what could potentially be included in data that is hashed

i will try to do some tests about it in my free time, i think that a score ID might be included there, since it's kinda an online identifier of the replay, but that's just speculation

Reply:
To be honest I think it'll be very hard to figure out what hash algorithm uses, especially because it might be intentionally obfuscated for anticheat purposes. Good luck though, and I'll be interested to hear if you do figure it out.